precompiles/pallet-democracy/src/lib.rs

```diff
@@ -0,0 +1,502 @@
+{
+	fn execute(
+		input: &[u8], //Reminder this is big-endian
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> Result<PrecompileOutput, ExitError> {
+		log::trace!(target: "democracy-precompile", "In democracy wrapper");
+
+		let mut input = EvmDataReader::new(input);
+
+		// Parse the function selector
+		// These are the four-byte function selectors calculated from the DemocracyInterface.sol
+		// according to the solidity specification
+		// https://docs.soliditylang.org/en/v0.8.0/abi-spec.html#function-selector
+		match &input.read_selector()? {
+			// Storage Accessors
+			Action::PublicPropCount => Self::public_prop_count(target_gas),
+			Action::DepositOf => Self::deposit_of(input, target_gas),
+			Action::LowestUnbaked => Self::lowest_unbaked(target_gas),
+			Action::OngoingReferendumInfo => Self::ongoing_referendum_info(input, target_gas),
+			Action::FinishedReferendumInfo => Self::finished_referendum_info(input, target_gas),
+
+			// Dispatchables
+			Action::Propose => Self::propose(input, target_gas, context),
+			Action::Second => Self::second(input, target_gas, context),
+			Action::StandardVote => Self::standard_vote(input, target_gas, context),
+			Action::RemoveVote => Self::remove_vote(input, target_gas, context),
+			Action::Delegate => Self::delegate(input, target_gas, context),
+			Action::UnDelegate => Self::un_delegate(target_gas, context),
+			Action::Unlock => Self::unlock(input, target_gas, context),
+		}
+	}
+}
+
+impl<Runtime> DemocracyWrapper<Runtime>
+where
+	Runtime: pallet_democracy::Config + pallet_evm::Config + frame_system::Config,
+	BalanceOf<Runtime>: TryFrom<U256> + TryInto<u128> + Debug + EvmData,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	Runtime::Call: From<DemocracyCall<Runtime>>,
+	Runtime::Hash: From<H256>,
+{
+	// The accessors are first. They directly return their result.
+
+	fn public_prop_count(target_gas: Option<u64>) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+
+		// Fetch data from pallet
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let prop_count = DemocracyOf::<Runtime>::public_prop_count();
+		log::trace!(target: "democracy-precompile", "Prop count from pallet is {:?}", prop_count);
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(prop_count).build(),
+			logs: Default::default(),
+		})
+	}
+
+	fn deposit_of(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+
+		// Bound check
+		input.expect_arguments(1)?;
+		let prop_index: u32 = input.read()?;
+
+		// Fetch data from pallet
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let deposit = match DemocracyOf::<Runtime>::deposit_of(prop_index) {
```

Comment:
```suggestion
		let deposit = DemocracyOf::<Runtime>::deposit_of(prop_index).ok_or(error("No such proposal in pallet democracy"))?;
```

Reply:
I chose an explicit `match` here because I thought `Some((_, deposit)) => deposit,` made it easier to notice that we aren't just returning the storage value here, but just part of the tuple. The suggested code returns the entire tuple. I've added a `.1` to the suggested code, but IMO that makes it less readable.

Reply:
Good point, I encourage keeping the most readable form, even if it is heavier.

Reply:
I've also used `ok_or_else` instead of `ok_or` because of https://rust-lang.github.io/rust-clippy/stable/index.html#or_fun_call.